### DISCUSSION TITLE
Broken recipe packet?

### DISCUSSION THREAD

**Discussion by BitCrack** [Minestom/Minestom#636] [Help] [answered]
I tried creating some recipes, I know all the values are correct, not null, etc.
![image](https://user-images.githubusercontent.com/19538752/152137422-0d8fbfad-508b-46ed-98f3-9aa8558fdcc4.png)
Minimal code:
```java
var resultId = NamespaceID.from("leather_helmet");
var resultMaterial = Material.fromNamespaceId(resultId);
var resultItem = ItemStack.of(resultMaterial);

var ingredientId = NamespaceID.from("leather");
var ingredientMaterial = Material.fromNamespaceId(ingredientId);
var ingredientItem = ItemStack.of(ingredientMaterial);

var e = new DeclareRecipesPacket.Ingredient(Collections.emptyList());
var i = new DeclareRecipesPacket.Ingredient(Collections.singletonList(ingredientItem));

var recipe = new ShapedRecipe(resultId.asString(), 3, 2, "", List.of(
    i, i, i,
    i, e, i
), resultItem) {
    @Override
    public boolean shouldShow(@NotNull Player player) {
        return true;
    }
};

MinecraftServer.getRecipeManager().addRecipe(recipe);
```
Though when I joined the server I got this exception:
![image](https://user-images.githubusercontent.com/19538752/152112707-18045c3e-1707-488c-9647-07e1dc0ef1fd.png)
I used Pakkit to check the packet and as I expected, it was malformed:
```
{
    recipes: [
        0: {
            type: "crafting_shaped",
            recipeId: "minecraft:turtle_helmet"
        },
        1: {
            type: "",
            recipeId: ""
        },
        2: {
            type: "",
            recipeId: ""
        },
        3: {
            type: "",
            recipeId: ""
        },
        4: {
            type: "�",
            recipeId: "cr"
        },
        5: {
            type: "fting_shapedminecraft:leather_helmet������crafting_shapedmine",
            recipeId: "raft:golden_helmet������crafting_shapedminecraft:iron_helmet"
        }
    ]
}
```

Also, I went ahead and tested Pakkit on other servers to see if it wasn't faulty, but it worked fine.
I tested this using Pakkit GitHub build 42 and Minestom b5cc2c4126 running behind a Velocity proxy.

Could there be a chance that empty ingredients are causing a problem? (Like seen in the first picture from the debugger, the turtle helmet doesn't have any ingredients because I haven't taken it into consideration yet and that's where the packet is starting to break, but I'm not sure)
And if it is, is this something a developer has to keep in mind, or should this be fixed in Minestom?

**Comment by mworzala** [comment]
Seems like it is related to empty ingredients. Passing air fixes it, though it is surprising the client can't handle an empty array there.

I would like to confirm in the client before adding a check to disallow an empty list.

(also, your reproducer is pretty bad. It has a number or errors)

**Reply by BitCrack** [reply]
Oh I see, I modified the code on GitHub and forgot that I might do it wrong. The reproducer should be good now.
Also I believe the exception might've been caused by an older Minestom version, when finishing writing this I decided to specify the version I used and got an idea to try the latest version and I noticed the exception was gone when connecting without Pakkit and I noticed that the code for this packet changed between these versions, but I haven't thought about it until now that you mentioned it.

But either way, I did pass air instead of an empty ingredient, but it still hasn't fixed it. I noticed the packet is still malformed in Pakkit and I still don't have the recipes in the recipe book and I think they should've shown up after the `unlock_recipes` packet. The packet has changed a bit though and this is how it looks in Pakkit now:
```
{
    recipes: [
        0: {
            type: "crafting_shaped",
            recipeId: "minecraft:turtle_helmet"
        },
        1: {
            type: "",
            recipeId: ""
        },
        2: {
            type: "",
            recipeId: ""
        },
        3: {
            type: "",
            recipeId: ""
        },
        4: {
            type: "",
            recipeId: "�"
        },
        5: {
            type: "cr",
            recipeId: "fting_shapedminecraft:leather_helmet������crafting_shapedmin"
        }
    ]
}
```

Also this is the `unlock_recipes`:
```
{
    action: 0,
    craftingBookOpen: false,
    filteringCraftable: false,
    smeltingBookOpen: false,
    filteringSmeltable: false,
    blastFurnaceOpen: false,
    filteringBlastFurnace: false,
    smokerBookOpen: false,
    filteringSmoker: false,
    recipes1: [
        0: "minecraft:turtle_helmet",
        1: "minecraft:leather_helmet",
        2: "minecraft:golden_helmet",
        3: "minecraft:iron_helmet",
        4: "minecraft:diamond_helmet",
        5: "minecraft:netherite_helmet"
    ],
    recipes2: [
        0: "minecraft:turtle_helmet",
        1: "minecraft:leather_helmet",
        2: "minecraft:golden_helmet",
        3: "minecraft:iron_helmet",
        4: "minecraft:diamond_helmet",
        5: "minecraft:netherite_helmet"
    ]
}
```

The only issue I can think of right now is sending *just* air for ingredients is causing issues, but I don't think it should? If Pakkit is reading packets just like the vanilla client then I don't see a reason why air instead of any other item would affect the packet in a significant way if they're serialized just like any other item?

**Reply by mworzala** [reply]
According to a cursory glance at minestom and vanilla 1.18 they seem to be doing the same thing. My guess would be that pakkit is doing something wrong and the client doesn't like recipes without any ingredients, but my guess is based on very little evidence.

It is rather confusing that pakkit thinks there are 6 recipes.

**Reply by BitCrack** [reply]
Oh, I forgot to mention, there are in fact 6 recipes, sorry for any trouble. Using the minimal code repeated 6 times for each type of helmet with ingredients matching the required ones for that helmet type (other than the turtle helmet for which I used empty and also air ingredients).

The reason why I suspect it might be a Minestom issue is that Pakkit is able to interpret the packet just fine on a server like Hypixel, e.g:
![image](https://user-images.githubusercontent.com/19538752/152302391-55a8ff3e-abb9-488c-8c3e-5f34dbc9814b.png)

Also, I tried also removing the turtle helmet that has just the air ingredients, and still, Pakkit has issues interpreting the packet (this time the packet doesn't show up at all?) and the helmets don't appear in the recipe book.

**Comment by BitCrack** [comment, marked as answer]
I'm sorry for the trouble, but Pakkit indeed interpreted the packet wrong. Weirdly though it seemed to do that on my Minestom server, but not e.g. Hypixel, which is also what threw me off in the first place.

It was my mistake because I wasn't aware of how the recipe book worked (In my defense I barely use it 😅) and just after debugging Minecraft I figured out that the recipe book doesn't show recipes for inventories that can't fit that recipe (I was checking for a 3x2 recipe in a 2x2 player crafting inventory). After quickly setting up the crafting inventory and opening it I indeed saw the recipes and also I went back and tried creating the recipe with empty ingredients (not air) and they appeared too.
